components/brave_wallet/browser/bitcoin/bitcoin_wallet_service.cc

```diff
@@ -0,0 +1,647 @@
+  std::vector<uint8_t> data;
+  for (auto& output : context.outputs) {
+    Push64AsLE(output.amount, data);
+    PushVarInt(output.script_pubkey.size(), data);
+    data.insert(data.end(), output.script_pubkey.begin(),
+                output.script_pubkey.end());
+  }
+  LOG(ERROR) << base::HexEncode(data);
+
+  auto double_hash = DoubleSHA256Hash(data);
+  to.insert(to.end(), double_hash.begin(), double_hash.end());
+}
+
+bool BitcoinWalletService::FillSignature(SendToContext& context,
+                                         uint32_t input_index) {
+  CHECK_LT(input_index, context.inputs.size());
+
+  auto& input = context.inputs[input_index];
+
+  // https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#specification
+  std::vector<uint8_t> data;
+  Push32AsLE(2, data);              // 1.
+  PushHashPrevouts(context, data);  // 2.
+  PushHashSequence(context, data);  // 3.
+
+  PushOutpoint(input.prev_output.outpoint, data);  // 4
+  PushScriptCode(context, input_index, data);      // 5.
+  Push64AsLE(input.prev_output.value, data);       // 6.
+  Push32AsLE(input.n_sequence, data);              // 7.
+
+  PushHashOutputs(context, data);      // 8.
+  Push32AsLE(context.locktime, data);  // 9.
+  Push32AsLE(kSigHashAll, data);       // 10.
+
+  if (auto signature = keyring_service_->SignMessageByBitcoinKeyring(
+          context.keyring_id, *input.key_id, DoubleSHA256Hash(data))) {
+    input.signature = std::move(*signature);
+  } else {
+    return false;
+  }
+
+  Push8AsLE(kSigHashAll, input.signature);
+  return true;
+}
+
+bool BitcoinWalletService::FillSignatures(SendToContext& context) {
+  for (uint32_t i = 0; i < context.inputs.size(); ++i) {
+    if (!FillSignature(context, i)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void PushInputs(SendToContext& context, std::vector<uint8_t>& to) {
+  PushVarInt(context.inputs.size(), to);
+  for (auto& input : context.inputs) {
+    PushOutpoint(input.prev_output.outpoint, to);
+    // TODO(apaymsyhev): support script for non-segwit transactions.
+    PushVarInt(0, to);
+    Push32AsLE(input.n_sequence, to);
+  }
+}
+
+void PushOutputs(SendToContext& context, std::vector<uint8_t>& to) {
+  PushVarInt(context.outputs.size(), to);
+  for (auto& output : context.outputs) {
+    Push64AsLE(output.amount, to);
+    PushVarInt(output.script_pubkey.size(), to);
+    to.insert(to.end(), output.script_pubkey.begin(),
+              output.script_pubkey.end());
+  }
+}
+
+void PushWitnesses(SendToContext& context, std::vector<uint8_t>& to) {
+  for (auto& input : context.inputs) {
+    // TODO(apaymyshev): only supports P2WPKH. Should support everything else.
+    PushVarInt(2, to);
+
+    PushVarSizeVector(input.signature, to);
+    PushVarSizeVector(input.pubkey, to);
+  }
+}
+
+bool BitcoinWalletService::FillTransaction(SendToContext& context) {
+  std::vector<uint8_t> data;
+  Push32AsLE(2, data);  // version
```

Comment:
Please add https://github.com/bitcoin/bips/blob/master/bip-0144.mediawiki#specification to the comment

Reply:
https://github.com/brave/brave-core/pull/17978/commits/c697f80928661571502650868b76f2f6252f298b